docs/MD_SEARCH.md

```diff
@@ -0,0 +1,209 @@
+--create-bucket-configuration LocationConstraint='us-east-1'
+# populate bucket
+aws --endpoint http://127.0.0.1:8000 s3api put-object--bucket bucketname \
+--key blueObject --metadata "color=blue" --tagging "type=color"
+aws --endpoint http://127.0.0.1:8000 s3api put-object--bucket bucketname \
+--key redObject --metadata "color=red" --tagging "type=color"
+
+./bin/search_bucket.js -a accessKey1 -k verySecretKey1 -b bucketname \
+-q "x-amz-meta-color=blue" -h 127.0.0.1 -p 8000
+```
+
+The above example describes the usage of the `search_bucket.js` script for
+issuing MD search request. This script can be found in the project root folder
+under the directory `bin`.
+
+#### Differences from SQL
+
+MD Search queries are very similar to the `WHERE` clauses of SQL queries, but
+they differ in that:
+
++ MD Searches queries follow the `PCRE` format
++ MD Search queries do not require dashed names, ex. `example-name`, to be
+  enclosed in backticks, ``(`)``
++ MD Search queries do not support all of the SQL operators.
+
+Supported SQL Operators: `=`, `<`, `>`, `<=`, `>=`, `!=`, `AND`, `OR`, `LIKE`,
+`<>`
+
+Unsupported SQL Operators:`NOT`, `BETWEEN`, `IN`, `IS`, `+`, `-`, `%`, `^`,
+`/`, `*`, `!`
+
+##### Using Regular Expressions in MD Search
+
+Regular expressions in MD search differs from SQL in that wildcards are
+represented with `.*` instead of `%`. The regex patterns must be wrapped in
+quotes as not doing so can lead to misinterpretation of the pattern.
+The regex patterns can be written in form of the `/pattern/` syntax or
+just the pattern if one does not require regex options, similar to `PCRE`.
+
+The supported regex flags are limited to the following to be compatible with
+mongodb:
+
++ `i`: case-insensitive
++ `m`: multi-line
++ `x`: extended
++ `s`: allow `.` to match newline
+
+Regex examples
+
+```
+# search for strings containing word substring "helloworld"
+    ".*helloworld.*"
+    "/.*helloworld.*/"
+    "/.*helloworld.*/i"
+```
+
+##### Specifying Metadata Fields
+
+To search common metadata headers:
+
+```
+    {metadata-key}{supported SQL op}{search value}
+    # example
+    key = blueObject
+    size > 0
+    key LIKE "blue.*"
+```
+
+To search custom user metadata:
+
+```
+    # metadata must be prefixed with "x-amz-meta-"
+    x-amz-meta-{usermetadata-key}{supported SQL op}{search value}
+    # example
+    x-amz-meta-color = blue
+    x-amz-meta-color != red
+    x-amz-meta-color LIKE "b.*"
+```
+
+To search tags:
+
+```
+    # tag searches must be prefixed with "tags."
+    tags.{tag-key}{supported SQL op}{search value}
+    # example
+    tags.type = color
+```
+
+#### Coding Examples
+
+The MD search requests can also be performed by making similar http request to
+example request below to the S3 Connector. The request must also be
+authenticated with the `AWS Signature version 4` method.
+
+```
+    GET /bucketname?search=hello%3Dworld HTTP/1.1
```

Comment:
Please show an example where you are searching user metadata or tags. It's more realworld like.